ERCS/erc-7820.md

```diff
@@ -0,0 +1,411 @@
+
+## Specification
+
+The `AccessControlRegistry` contract provides a standardized interface for managing access control in Ethereum smart contracts. It includes functions to register and unregister contracts, grant and revoke roles for specific contracts, and check if an account has a particular role in a registered contract. Events are emitted for contract registration, unregistration, role grants, and role revocations, ensuring transparency and traceability of access control changes.
+
+Additionally, the AccessControlRegistry MUST reject the registration of zero addresses.
+
+```solidity
+pragma solidity 0.8.23;
+
+interface IAccessControlRegistry {
+
+    // Contains information about a registered contract.
+    // @param isActive Indicates whether the contract is active.
+    // @param admin The address of the admin for the registered contract.
+    struct ContractInfo {
+        bool isActive;
+        address admin;
+    }
+
+    // Emitted when a contract is registered.
+    // @param _contract The address of the registered contract.
+    // @param _admin The address of the admin for the registered contract.
+    event ContractRegistered(address indexed _contract, address indexed _admin);
+
+    // Emitted when a contract is unregistered.
+    // @param _contract The address of the unregistered contract.
+    // @param _admin The address who unregistered the contract
+    event ContractUnregistered(address indexed _contract, address indexed _admin);
+
+    // Emitted when a role is granted to an account for a contract.
+    // @param targetContract The address of the contract.
+    // @param role The role being granted.
+    // @param account The address of the account.
+    event RoleGranted(
+        address indexed targetContract,
+        bytes32 indexed role,
+        address indexed account
+    );
+
+    // Emitted when a role is revoked from an account for a contract.
+    // @param targetContract The address of the contract.
+    // @param role The role being revoked.
+    // @param account The address of the account.
+    event RoleRevoked(
+        address indexed targetContract,
+        bytes32 indexed role,
+        address indexed account
+    );
+
+    // Registers a contract with the given admin.
+    // @param _admin The address of the admin for the registered contract.
+    function registerContract(address _admin) external;
+
+    // Unregisters a contract.
+    // @param _contract The address of the contract to unregister.
+    function unRegisterContract(address _contract) external;
+
+    // Grants roles to multiple accounts for multiple contracts.
+    // @param targetContracts An array of contract addresses to which roles will be granted.
+    // @param roles An array of roles to be granted.
+    // @param accounts An array of accounts to be granted the roles.
+    function grantRole(
+        address[] memory targetContracts,
+        bytes32[] memory roles,
+        address[] memory accounts
+    ) external;
```

Comment:
The documentation for this function is a bit ambiguous. Does this grant all the listed `roles` to all the `targetContracts` for all the `accounts`, or is it more like a tuple `(contract, role, account)`?